# Wrestle the Android App Store Data into Beautiful Looking Charts with Plotly

Have you ever thought about building your own an iOS or Android app? If so, then you probably have wondered about how things work in the app stores. Today we'll replicate some of the app store analytics provided by companies like App Annie or Sensor Tower that helps inform development and app marketing strategies for many companies. This stuff is BIG business!

We will compare thousands of apps in the Google Play Store so that we can gain insight into:

- How competitive different app categories (e.g., Games, Lifestyle, Weather) are

- Which app category offers compelling opportunities based on its popularity

- How many downloads you would give up by making your app paid vs. free

- How much you can reasonably charge for a paid app

- Which paid apps have had the highest revenue

- How many paid apps will recoup their development costs based on their sales revenue



In [1]:
import pandas as pd

df_apps = pd.read_csv('apps.csv')

# Data Cleaning: Removin NaN Values and Duplicates

The first step as always is getting a better idea about what we're dealing with.

### Preliminary Data Exploration

How many rows and columns does df_apps have? What are the column names? What does the data look like? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html)

In [ ]:
# How many rows and columns does df_apps have?
df_apps.shape

# 10841 rows, 12 columns

(10841, 12)

We are working with a fairly large DataFrame this time.

10841 rows and 12 columns.

In [11]:
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0       8.7        0  Paid  $13.99           Teen           Social   
1        0      33.0        0  Paid   $2.99       Everyone        Education   
2        0       5.5        0  Paid   $1.49       Everyone  Personalization   
3        0      19.0        0   NaN       0   Everyone 10+         Strategy   
4        0      10.0        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

We can already see that there are some data issues that we need to fix. In the Ratings and Type columns there are NaN (Not a number values) and in the Price column we have dollar signs that will cause problems.



In [ ]:
# What are the column names? 
print('Columns: ')
for col in df_apps.columns:
    print(f' - {col}')

Columns: 
 - App
 - Category
 - Rating
 - Reviews
 - Size_MBs
 - Installs
 - Type
 - Price
 - Content_Rating
 - Genres
 - Last_Updated
 - Android_Ver


In [ ]:
# What does the data look like? Look at a random sample of 5 different rows with .sample()
df_apps.sample(5)

App       Category  \
6002  Chatting - Free chat, random chat, boyfriend, ...         DATING   
1433                                   CW Bluetooth SPP  COMMUNICATION   
6747                              Go-Go-Goat! Free Game         FAMILY   
5855                 Zombie War Z : Hero Survival Rules           GAME   
3084                        Check Your Visitors on FB ?         SOCIAL   

      Rating  Reviews   Size_MBs   Installs  Type Price Content_Rating  \
6002     4.2     2506   6.100000    500,000  Free     0     Mature 17+   
1433     NaN        3   0.938477        100  Free     0       Everyone   
6747     4.5    66740  23.000000  1,000,000  Free     0       Everyone   
5855     3.9     1987  62.000000    100,000  Free     0     Mature 17+   
3084     3.6       40   1.500000      5,000  Free     0       Everyone   

                         Genres       Last_Updated   Android_Ver  
6002                     Dating      June 15, 2017    4.0 and up  
1433              Communication     March 31, 2018    6.0 and up  
6747  Arcade;Action & Adventure       July 8, 2016    2.3 and up  
5855                     Action  December 13, 2017  4.0.3 and up  
3084                     Social  February 17, 2017    4.2 and up

Remove the columns called Last_Updated and Android_Version from the DataFrame. We will not use these columns.



In [12]:
df_apps = df_apps.drop('Last_Updated', axis=1) #axis = 0 for rows, axis = 1 for columns
df_apps = df_apps.drop('Android_Ver', axis=1)

In [13]:
print('Columns: ')
for col in df_apps.columns:
    print(f' - {col}')

Columns: 
 - App
 - Category
 - Rating
 - Reviews
 - Size_MBs
 - Installs
 - Type
 - Price
 - Content_Rating
 - Genres


How many rows have a NaN value (not-a-number) in the Rating column? Create DataFrame called df_apps_clean that does not include these rows.

In [16]:
df_apps.Rating.isna().sum()

np.int64(1474)

In [24]:
# df_apps_clean = df_apps[df_apps.Rating.notna()]
df_apps_clean = df_apps.dropna()

In [26]:
df_apps_clean.shape

(9367, 10)

This leaves us with 9,367 entries in our DataFrame. But there may be other problems with the data too:

Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.



In [33]:
df_apps_clean[df_apps_clean.duplicated()]

App            Category  Rating  \
946                        420 BZ Budeze Delivery             MEDICAL     5.0   
1133                                  MouseMingle              DATING     2.7   
1196   Cardiac diagnosis (heart rate, arrhythmia)             MEDICAL     4.4   
1231                                 Sway Medical             MEDICAL     5.0   
1247               Chat Kids - Chat Room For Kids              DATING     4.7   
...                                           ...                 ...     ...   
10802               Skype - free IM & video calls       COMMUNICATION     4.1   
10809                                   Instagram              SOCIAL     4.5   
10826                                Google Drive        PRODUCTIVITY     4.4   
10832                                 Google News  NEWS_AND_MAGAZINES     3.9   
10839                              Subway Surfers                GAME     4.5   

        Reviews  Size_MBs       Installs  Type   Price Content_Rating  \
946           2      11.0            100  Free       0     Mature 17+   
1133          3       3.9            100  Free       0     Mature 17+   
1196          8       6.5            100  Paid  $12.99       Everyone   
1231          3      22.0            100  Free       0       Everyone   
1247          6       4.9            100  Free       0     Mature 17+   
...         ...       ...            ...   ...     ...            ...   
10802  10484169       3.5  1,000,000,000  Free       0       Everyone   
10809  66577313       5.3  1,000,000,000  Free       0           Teen   
10826   2731211       4.0  1,000,000,000  Free       0       Everyone   
10832    877635      13.0  1,000,000,000  Free       0           Teen   
10839  27725352      76.0  1,000,000,000  Free       0   Everyone 10+   

                 Genres  
946             Medical  
1133             Dating  
1196            Medical  
1231            Medical  
1247             Dating  
...                 ...  
10802     Communication  
10809            Social  
10826      Productivity  
10832  News & Magazines  
10839            Arcade  

[476 rows x 10 columns]

In [36]:
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL     4.5  66577313       5.3  1,000,000,000  Free   
10808  Instagram   SOCIAL     4.5  66577446       5.3  1,000,000,000  Free   
10809  Instagram   SOCIAL     4.5  66577313       5.3  1,000,000,000  Free   
10810  Instagram   SOCIAL     4.5  66509917       5.3  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10809     0           Teen  Social  
10810     0           Teen  Social

In [38]:
df_apps_clean[df_apps_clean.App == 'Instagram'].duplicated()

10806    False
10808    False
10809     True
10810    False
dtype: bool

In [39]:
df_apps_clean = df_apps_clean.drop_duplicates()

In [40]:
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL     4.5  66577313       5.3  1,000,000,000  Free   
10808  Instagram   SOCIAL     4.5  66577446       5.3  1,000,000,000  Free   
10810  Instagram   SOCIAL     4.5  66509917       5.3  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10810     0           Teen  Social

We need to provide the column names that should be used in the comparison to identify duplicates. 

In [41]:
df_apps_clean = df_apps_clean.drop_duplicates(subset=['App', 'Type', 'Price'])

In [42]:
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL     4.5  66577313       5.3  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social

In [43]:
df_apps_clean.shape

(8199, 10)

This leaves us with 8,199 entries after removing duplicates. 

### What else should I know about the data?

So we can see that 13 different features were originally scraped from the Google Play Store.

- Obviously, the data is just a sample out of all the Android apps. It doesn't include all Android apps of which there are millions.

- I’ll assume that the sample is representative of the App Store as a whole. This is not necessarily the case as, during the web scraping process, this sample was served up based on geographical location and user behaviour of the person who scraped it - in our case Lavanya Gupta.

- The data was compiled around 2017/2018. The pricing data reflect the price in USD Dollars at the time of scraping. (developers can offer promotions and change their app’s pricing).

- I’ve converted the app’s size to a floating-point number in MBs. If data was missing, it has been replaced by the average size for that category.

- The installs are not the exact number of installs. If an app has 245,239 installs then Google will simply report an order of magnitude like 100,000+. I’ve removed the '+' and we’ll assume the exact number of installs in that column for simplicity.

